File: web/lib/indicators.ts
import type { DailyPrice, MacdSignalType, TechnicalIndicator } from "./types";

export interface MACDPoint {
  date: string;
  dif: number;
  signal: number;
  histogram: number;
}

function round(value: number): number {
  return Math.round(value * 10000) / 10000;
}

export function calculateEMA(values: number[], period: number): number[] {
  if (!values.length) return [];
  const multiplier = 2 / (period + 1);
  const result = [values[0]];
  for (let index = 1; index < values.length; index += 1) {
    result.push(values[index] * multiplier + result[index - 1] * (1 - multiplier));
  }
  return result;
}

export function calculateMACD(
  prices: Pick<DailyPrice, "date" | "close">[],
  fastPeriod = 12,
  slowPeriod = 26,
  signalPeriod = 9,
): MACDPoint[] {
  if (!prices.length) return [];
  const closes = prices.map((price) => price.close);
  const fast = calculateEMA(closes, fastPeriod);
  const slow = calculateEMA(closes, slowPeriod);
  const dif = fast.map((value, index) => value - slow[index]);
  const signal = calculateEMA(dif, signalPeriod);
  return prices.map((price, index) => ({
    date: price.date,
    dif: round(dif[index]),
    signal: round(signal[index]),
    histogram: round(dif[index] - signal[index]),
  }));
}

export function generateMACDSignals(
  macd: Pick<MACDPoint, "date" | "histogram">[],
): { date: string; macdSignal: MacdSignalType }[] {
  return macd.map((point, index) => {
    if (index === 0) return { date: point.date, macdSignal: null };
    const previous = macd[index - 1].histogram;
    const current = point.histogram;
    if (previous < 0 && current >= 0) return { date: point.date, macdSignal: "entry" };
    if (previous >= 0 && current < 0) return { date: point.date, macdSignal: "exit" };
    return { date: point.date, macdSignal: null };
  });
}

export function calculateSMA(values: number[], period: number): (number | null)[] {
  let rolling = 0;
  return values.map((value, index) => {
    rolling += value;
    if (index >= period) rolling -= values[index - period];
    return index >= period - 1 ? round(rolling / period) : null;
  });
}

export function calculateIndicators(prices: DailyPrice[]): TechnicalIndicator[] {
  const closes = prices.map((price) => price.close);
  const periods = [5, 10, 20, 30, 60, 120, 240] as const;
  const averages = Object.fromEntries(
    periods.map((period) => [period, calculateSMA(closes, period)]),
  ) as Record<(typeof periods)[number], (number | null)[]>;
  const macd = calculateMACD(prices);
  const signals = generateMACDSignals(macd);

  return prices.map((price, index) => ({
    date: price.date,
    ma5: averages[5][index],
    ma10: averages[10][index],
    ma20: averages[20][index],
    ma30: averages[30][index],
    ma60: averages[60][index],
    ma120: averages[120][index],
    ma240: averages[240][index],
    dif: macd[index]?.dif ?? null,
    signal: macd[index]?.signal ?? null,
    histogram: macd[index]?.histogram ?? null,
    macdSignal: signals[index]?.macdSignal ?? null,
  }));
}

File: backend/app/services/indicators.py
from collections.abc import Iterable


def moving_average(values: list[float], period: int) -> list[float | None]:
    result: list[float | None] = []
    running = 0.0
    for index, value in enumerate(values):
        running += value
        if index >= period:
            running -= values[index - period]
        result.append(round(running / period, 4) if index >= period - 1 else None)
    return result


def ema(values: Iterable[float], period: int) -> list[float]:
    items = list(values)
    if not items:
        return []
    multiplier = 2 / (period + 1)
    result = [items[0]]
    for value in items[1:]:
        result.append(value * multiplier + result[-1] * (1 - multiplier))
    return result


def calculate_macd(closes: list[float]) -> list[dict[str, float | None]]:
    fast = ema(closes, 12)
    slow = ema(closes, 26)
    dif = [left - right for left, right in zip(fast, slow, strict=True)]
    signal = ema(dif, 9)
    return [
        {
            "dif": round(dif[index], 4),
            "signal": round(signal[index], 4),
            "histogram": round(dif[index] - signal[index], 4),
        }
        for index in range(len(closes))
    ]


def generate_macd_signals(macd: list[dict[str, float | None]]) -> list[str | None]:
    result: list[str | None] = [None] * len(macd)
    for index in range(1, len(macd)):
        previous = macd[index - 1]["histogram"]
        current = macd[index]["histogram"]
        dif = macd[index].get("dif")
        signal = macd[index].get("signal")
        if previous is None or current is None or dif is None or signal is None:
            continue
        if dif <= 0 or signal <= 0:
            continue
        if previous < 0 <= current:
            result[index] = "entry"
        elif previous >= 0 > current:
            result[index] = "exit"
    return result


def calculate_indicators(candles: list[dict]) -> list[dict]:
    closes = [float(item["close"]) for item in candles]
    averages = {period: moving_average(closes, period) for period in (5, 10, 20, 30, 60, 120, 240)}
    macd = calculate_macd(closes)
    signals = generate_macd_signals(macd)
    return [
        {
            "date": candle["date"],
            **{f"ma{period}": averages[period][index] for period in averages},
            **macd[index],
            "macdSignal": signals[index],
        }
        for index, candle in enumerate(candles)
    ]
